chalk-derive/src/lib.rs

```diff
@@ -147,6 +147,140 @@ pub fn derive_fold(item: TokenStream) -> TokenStream {
     panic!("derive(Fold) requires a parameter that implements HasTypeFamily or TypeFamily");
 }
 
+/// Derives Fold for structs and enums for which one of the following is true:
+/// - It has a `#[has_type_family(TheTypeFamily)]` attribute
+/// - There is a single parameter `T: HasTypeFamily` (does not have to be named `T`)
+/// - There is a single parameter `TF: TypeFamily` (does not have to be named `TF`)
+#[proc_macro_derive(SuperFold, attributes(has_type_family))]
+pub fn derive_super_fold(item: TokenStream) -> TokenStream {
```

Comment:
So I was looking at this again. The only place this is used is on `ProgramClause`. Do you think this will get used in the future elsewhere? Tbh, it's a lot of complexity for a single derive.

Reply:
Hmm, good point. I think I originally expected to use it on `Goal` and `DomainGoal` but for various reasons didn't wind up doing so. One reason in particular that is pretty broadly applicable is that the `Goal<'tcx>` / `GoalData<'tcx>` split means that "super fold" on `Goal` is effectively just *fold* on `GoalData`. I think we could remove it.

it's also true that this code is largely copy and paste from the other fold definition, so it could probably be collapsed.

Reply:
Removed the `SuperFold` derive.